packages/@react-stately/color/src/useColorFieldState.ts

```diff
@@ -71,40 +71,41 @@ export function useColorFieldState(
 
   let initialValue = useColor(value);
   let initialDefaultValue = useColor(defaultValue);
-  let [colorValue, setColorValue] = useControlledState<Color>(initialValue, initialDefaultValue, onChange);
-  let [inputValue, setInputValue] = useState(() => (value || defaultValue) && colorValue ? colorValue.toString('hex') : '');
+  let [colorValue, setColorValue] = useControlledState<Color | null>(initialValue, initialDefaultValue, onChange);
+  let [inputValue, _setInputValue] = useState(() => (value || defaultValue) && colorValue ? colorValue.toString('hex') : '');
+  let [parsedColor, setParsedColor] = useState<Color | null>(colorValue);
+  let [forceSync, setForceSync] = useState(false);
 
-  let safelySetColorValue = (newColor: Color) => {
-    if (!colorValue || !newColor) {
-      setColorValue(newColor);
-      return;
-    }
-    if (newColor.toHexInt() !== colorValue.toHexInt()) {
-      setColorValue(newColor);
-      return;
-    }
-  };
-
-  let prevValue = useRef(colorValue);
-  if (prevValue.current !== colorValue) {
-    setInputValue(colorValue ? colorValue.toString('hex') : '');
-    prevValue.current = colorValue;
-  }
-
-
-  let parsedValue = useMemo(() => {
+  let setInputValue = useCallback((val) => {
+    _setInputValue(val);
     let color;
     try {
-      color = parseColor(inputValue.startsWith('#') ? inputValue : `#${inputValue}`);
+      color = parseColor(val.startsWith('#') ? val : `#${val}`);
     } catch (err) {
       color = null;
     }
-    return color;
-  }, [inputValue]);
-  let parsed = useRef(null);
-  parsed.current = parsedValue;
+    setParsedColor(color);
+  }, [_setInputValue, setParsedColor]);
+
+  let prevValue = useRef(colorValue);
+  useEffect(() => {
+    if (forceSync || prevValue.current !== colorValue) {
+      setInputValue(colorValue ? colorValue.toString('hex') : '');
+      setForceSync(false);
+      prevValue.current = colorValue;
+    }
+  }, [forceSync, colorValue, setInputValue]);
```

Comment:
There are a lot of changes here, I presume to deal with the useEffect being async? Do you remember? In #4564, I made `prevValue` stored in state instead of a ref so we can still update it in render.

Reply:
this had to do with syncing the input value to the state in a controlled component where the controlling value wasn't updated after onChange

it appears that using state to store the previous value works well though,